src/creeps/Zombie.cpp

```diff
@@ -1,76 +1,287 @@
+    // Frames are used to make sure the zombie doesn't move through the path too quickly/slowly
+    if (frame > 0) {
+        --frame;
+    } else {
+        setCurFrame(ZOMBIE_FRAMES);
+        ++step;
+    }
 
-    return distr(eng);
+    setCurDir(direction);
+    setState(ZombieState::ZOMBIE_MOVE);
 }
 
-void Zombie::generateRandomMove(){
-    double cosVal = cos(getAngle()*PI/180.0);
-    double sinVal = sin(getAngle()*PI/180.0);
+/**
+ * A* algo generates a string of direction digits.
+ * Fred Yang
+ * Feb 14
+ */
+string Zombie::generatePath(const float xStart, const float yStart,
+        const float xDest, const float yDest) {
+    // priority queue index
+    int index = 0;
 
-    const int x = getVelocity();
-    const int y = getVelocity();
+    // temp index
+    int i;
+    int j;
 
-    const int tx = getX() + 20 * cosVal;
-    const int ty = getY() + 20 * sinVal;
+    // row / column index
+    int x;
+    int y;
+    int xdx;
+    int ydy;
 
-    Entity::moveMoveHitBox(tx, ty);
-    Entity::moveProHitBox(tx,ty);
+    // path to be generated
+    string path;
 
-    CollisionHandler &ch = GameManager::instance()->getCollisionHandler();
+    // current node & child node
+    static Node curNode;
+    static Node childNode;
 
-    if(ch.detectMovementCollision(ch.getQuadTreeEntities(ch.quadtreeMarine,this),this)
-            || ch.detectMovementCollision(ch.getQuadTreeEntities(ch.quadtreeZombie,this),this)
-            || ch.detectMovementCollision(ch.getQuadTreeEntities(ch.quadtreeBarricade,this),this)
-            || ch.detectMovementCollision(ch.getQuadTreeEntities(ch.quadtreeWall,this),this)
-            || ch.detectMovementCollision(ch.getQuadTreeEntities(ch.quadtreeTurret,this),this)
-            || ch.detectMovementCollision(ch.getQuadTreeEntities(ch.quadtreeObj,this),this)) {
+    // priority queue
+    static array<priority_queue<Node>, 2> pq;
+    //static priority_queue<Node> pq[2];
 
-        setAngle(getRandomAngle());
-        cosVal = cos(getAngle() * PI / 180.0);
-        sinVal = sin(getAngle() * PI / 180.0);
-    }
+    // reset the node maps
+    memset(closedNodes, 0, sizeof(closedNodes[0][0]) * ROW * COL);
+    memset(openNodes, 0, sizeof(openNodes[0][0]) * ROW * COL);
+
+    const int xNodeStart = static_cast<int>(xStart / TILE_SIZE);
+    const int yNodeStart = static_cast<int>(yStart / TILE_SIZE);
+    const int xNodeDest = static_cast<int>(xDest / TILE_SIZE);
+    const int yNodeDest = static_cast<int>(yDest / TILE_SIZE);
+
+    // create the start node and push into open list
+    curNode = Node(xNodeStart, yNodeStart, 0, 0);
+    curNode.updatePriority(xNodeDest, yNodeDest);
+    pq[index].push(curNode);
+
+    // A* path finding
+    while (!pq[index].empty()) {
+        Node tmp = pq[index].top();
+        // get the current node with the highest priority from open list
+        curNode = Node(tmp.getXPos(), tmp.getYPos(), tmp.getLevel(), tmp.getPriority());
```

Comment:
Okay so this seems weird to me
You assign tmp to curNode, but instead of assigning directly, you create a temp node and assign that?
That's a waste
Plus it seems like you don't modify tmp or curNode for the rest of the method
I can understand using a temp to store a value that gets deleted or overwritten
But this scheme seems odd.
Firstly, curNode and childNode don't need to be static.
They are constantly reassigned, so making them static is confusing and unintuitive.
Secondly, delete tmp
You assign the temp value to curNode anyhow, so no need for the tmp variable
Lastly, just make curNode equal to `pq[index].top` and be done with it
No weird creating a temporary and assigning from that